### PR TITLE
VIH-7283 search judge accounts

### PR DESCRIPTION
**Pull request by FarronF**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/VIH-7283

### Change description ###
Allow search for judiciary and courtroom account judges.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.html [hunk 1]
```diff
@@ -4,6 +4,33 @@
...
+<div
+  *ngIf="errorAlternativeEmail || errorJohAccountNotFound || errorJudiciaryAccount"
+  class="govuk-error-summary govuk-!-width-full"
+  aria-labelledby="error-summary-title"
+  role="alert"
+  tabindex="-1"
+  data-module="error-summary"
+>
+  <h2 class="govuk-error-summary__title" id="error-summary-title">
+    {{ 'There is a problem' }}
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.html]: {{ }} aren't needed here. Only for interpolation.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts [hunk 2]
```diff
@@ -161,10 +163,22 @@ export class AddParticipantComponent extends BookingBaseComponent implements OnI
         if (this.editMode) {
             if (this.searchEmail && this.participantDetails) {
                 this.setParticipantEmail();
+                this.subcribeForSeachEmailEvents();
             }
         }
     }
 
+    subcribeForSeachEmailEvents() {
+        this.searchEmail.notFoundEmailEvent$.subscribe(notFound => {
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: I think the email component should maintain the error state itself - i.e., errorAlternativeEmail & errorJohAccountNotFound should exist in there and not in this component. This component should perhaps just assign a local variable to something like `this.searchEmail.errorAlternativeEmail$' and expose this to the UI.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts [hunk 3]
```diff
@@ -300,6 +314,13 @@ export class AddParticipantComponent extends BookingBaseComponent implements OnI
     }
 
     public getParticipant(participantDetails: ParticipantModel) {
+        if (!this.validateJudgeAndJohMembers()) {
+            this.searchEmail.isErrorEmailAssignedToJudge = true;
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: As per above, the searchEmail component maintains some of the state.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts [hunk 4]
```diff
@@ -496,9 +531,47 @@ export class AddParticipantComponent extends BookingBaseComponent implements OnI
...
+        if (this.hearing?.participants.length) {
+            const judge = this.hearing.participants.find(x => x.is_judge);
+            return this.searchEmail.email !== judge?.username;
+        }
+        return true;
+    }
+
+    validateJudiciaryEmailAndRole() {
+        if (this.searchEmail && this.searchEmail.email.length) {
+            this.searchService.searchJudiciaryEntries(this.searchEmail.email).subscribe(x => {
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: I think the searchEmail component should do this searching.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts [hunk 5]
```diff
@@ -496,9 +531,47 @@ export class AddParticipantComponent extends BookingBaseComponent implements OnI
...
+                this.errorJudiciaryAccount = false;
+                if (x && x.length) {
+                    if (this.role.value !== 'Panel Member' && this.role.value !== 'Winger') {
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: Both branches of this `if` call the same func. Is this right?

Comment (phillipjchambers) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: Also the nesting looks a bit odd

Comment (phillipjchambers) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: I _think_ I see what's happening, but perhaps we can make it a bit more readable.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html [hunk 6]
```diff
@@ -16,13 +16,13 @@ <h2 class="govuk-error-summary__title" id="error-summary-title">
   </h2>
   <div class="govuk-error-summary__body">
     <ul class="govuk-list govuk-error-summary__list">
-      <li *ngIf="judgeName.invalid">
-        <a (click)="goToDiv('judgeName-error')">Please select a courtroom account</a>
+      <li *ngIf="!judge">
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html:19]: Consider using templates here

Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html:19]: Consider using templates here

**Review by phillipjchambers — CHANGES_REQUESTED** [review]

File: AdminWebsite/AdminWebsite/ClientApp/src/app/services/video-hearings.service.ts [hunk 7]
```diff
@@ -428,4 +429,29 @@ export class VideoHearingsService {
     getTelephoneConferenceId(hearingId: string): Observable<PhoneConferenceResponse> {
         return this.bhClient.getTelephoneConferenceIdById(hearingId);
     }
+
+
+    canAddUser(username: string): boolean {
+        const userPresent = this.modelHearing.participants.find(x => x.username.toLowerCase() === username.toLowerCase());
+        if (!userPresent) {
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/services/video-hearings.service.ts:436]: You can simplify this:
```
return !userPresent;
```

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 8]
```diff
@@ -104,32 +107,62 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
...
+        this.canNavigate = false;
+        this.isJudgeParticipantError = false;
+        if (judge) {
+            if (this.hearingService.canAddJudge(judge.username)) {
+                judge.is_judge = true;
+                this.courtAccountJudgeEmail = judge.email;
+                this.judgeDisplayNameFld.setValue(this.judge.display_name);
+                this.hearing.participants = this.hearing.participants.filter(x => !x.is_judge);
+                this.hearing.participants.unshift(judge);
+                this.hearing = Object.assign({}, this.hearing);
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:147]: use a spread operator here, it's a bit neater & less verbose

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 9]
```diff
@@ -104,32 +107,62 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
...
+                this.canNavigate = true;
+            } else {
+                this.isJudgeParticipantError = true;
+            }
+        } else {
+            this.removeJudge();
+        }
+
+        this.isJudgeSelected = judge != null;
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:156]: !!judge maybe? What's this for? Maybe this one could just be a getter as it looks like it's connected to this.judge anyway?

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 10]
```diff
@@ -206,49 +218,37 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
...
-            return;
-        }
-        this.judge.first_name = selectedJudge.first_name;
-        this.judge.last_name = selectedJudge.last_name;
-        this.judge.email = selectedJudge.email;
-        this.courtAccountJudgeEmail = selectedJudge.email;
-        if (!this.isJudgeDisplayNameSet()) {
-            this.judge.display_name = selectedJudge.display_name;
+    get displayEmailField() {
+        if (this.judge != null && this.judge.is_courtroom_account) {
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:225]: `return !!this.judge && this.judge.is_courtroom_account` perhaps?

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 11]
```diff
@@ -206,49 +218,37 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
...
-        this.hearing.participants.unshift(newJudge);
-        this.hearing = Object.assign({}, this.hearing);
     }
 
     isJudgeDisplayNameSet(): boolean {
         let result = false;
         if (this.judge && this.judge.display_name) {
-            const enteredJudge = this.availableJudges.find(j => j.display_name === this.judge.display_name);
-            result = !enteredJudge;
+            result = true;
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:235]: I think you could one line this whole func now that it's simpler

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/search-email/search-email.component.spec.ts [hunk 12]
```diff
@@ -12,34 +12,52 @@ import { SearchEmailComponent } from './search-email.component';
...
-        "first_name": "Alisa",
-        "middle_names":"No",
-        "last_name": "Smith",
-        "telephone_number": "1111222222",
-        "username": "vb.email2@hmcts.net"
-      }
-    ]
-    `
-    );
+    // const participantList: PersonResponse[] = JSON.parse(
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/search-email/search-email.component.spec.ts:15]: Remove comments 👍

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/search-email/search-email.component.ts [hunk 13]
```diff
@@ -43,23 +48,37 @@ export class SearchEmailComponent implements OnInit, OnDestroy {
...
+                        this.searchPending.next(true);
+                    }),
+                    switchMap(term => {
+                        return this.searchService.participantSearch(term, this.hearingRoleParticipant);
+                    })
+                )
+
+                .subscribe(personsFound => {
+                    if (personsFound && personsFound.length > 0) {
+                        this.getData(personsFound);
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/search-email/search-email.component.ts:65]: Is there a reason getData isn't just in the pipe?

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/services/participant-mapper.service.ts [hunk 14]
```diff
@@ -0,0 +1,45 @@
+import { Injectable } from '@angular/core';
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/services/participant-mapper.service.ts:1]: Let's talk about this mapper, I'm not sure it's needed.

Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/services/participant-mapper.service.ts:1]: Yeah seems overkill. I was thinking to move these into PaticipantModel. Does that make sense to you?

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/search-email/search-email.component.ts (hunk 13 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/search-email/search-email.component.ts:65]: Not my change, I'll look into it.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts (hunk 8 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:147]: Agreed.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts (hunk 9 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:156]: Yeah there's a few of these types of things sprinkled around that I was hesitant to change. I think this one can be no problem

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts (hunk 11 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:235]: Agreed, thanks

File: AdminWebsite/AdminWebsite/ClientApp/src/app/services/video-hearings.service.ts (hunk 7 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/services/video-hearings.service.ts:436]: Yeah, not sure why I did it like this.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts (hunk 4 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/add-participant/add-participant.component.ts]: These changes are from VIH-7284 which is already merged into master. This branch was based on that branch before it was merged into master. I didn't expect commits from that branch to show up in this pull request.

This is called when the role is changed to ensure the user that was previously searched it is still valid. It probably should be refactored to go through the searchEmail component though.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts (hunk 8 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts]: Actually this line isn't needed here at all.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/services/participant-mapper.service.ts (hunk 14 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/services/participant-mapper.service.ts]: Removed the service but I kept the code as static methods in ParticipantModel.

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html (hunk 6 above)
Comment (FarronF) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html:19]: I was just using the structure that was there before. Templates could be a better fit if you want me to do it as part of this I could.

Comment (shaed-parkar) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html:19]: Feel free to make improvements to the code as you find issues like these

Comment (phillipjchambers) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.html:19]:
```html
<ng-container *ngIf="judge; else nonJudge">
      <li *ngIf="judgeDisplayNameInvalid">
        <a (click)="goToDiv('judgeDisplayName-error')">{{ constants.Error.JudgeDisplayNameErrorMsg }}</a>
      </li>
      <li *ngIf="isJudgeParticipantError">
        <a (click)="goToDiv('judgeDisplayName-error')">{{ constants.Error.JudgeParticipantErrorMsg }}</a>
      </li>
</ng-container>
<ng-template #nonJudge>
  <li>
    <a (click)="goToDiv('judgeEmail-error')">{{ constants.Error.JudgeEmailErrorMsg }}</a>
  </li>
</ng-template>
```

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 15]
```diff
@@ -186,8 +202,8 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
         return this.form.get('judgeName');
     }
 
-    get judgeNameInvalid() {
-        return this.judgeName.invalid && (this.judgeName.dirty || this.judgeName.touched || this.failedSubmission);
+    get isJudgeSelected() {
+        return !!this.judge && !!this.judge.email && this.judge.email !== '';
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:206]: `return !!this.judge && !!this.judge.email;`

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 16]
```diff
@@ -206,49 +222,29 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
...
     }
 
     isJudgeDisplayNameSet(): boolean {
-        let result = false;
-        if (this.judge && this.judge.display_name) {
-            const enteredJudge = this.availableJudges.find(j => j.display_name === this.judge.display_name);
-            result = !enteredJudge;
-        }
-        return result;
+        return !!this.judge && !!this.judge.display_name && this.judge.display_name !== '';
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:233]: As above

Comment (phillipjchambers) [reply AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:233]: return this.judge && this.judge.display_name

File: AdminWebsite/AdminWebsite/ClientApp/src/app/services/search.service.ts [hunk 17]
```diff
@@ -60,34 +63,56 @@ export class SearchService {
...
-        if (term.length > 2) {
+        if (term.length >= this.minimumSearchLength) {
             return this.bhClient.postJudiciaryPersonBySearchTerm(term);
+        } else {
+            return of(allResults);
+        }
+    }
+
+    searchJudgeAccounts(term): Observable<Array<JudgeResponse>> {
+        if (term.length >= this.minimumSearchLength) {
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/services/search.service.ts:112]: I would remove the check from here & just add it to the component (in the pipe use a filter)

File: AdminWebsite/AdminWebsite/ClientApp/src/app/services/video-hearings.service.ts [hunk 18]
```diff
@@ -428,4 +430,20 @@ export class VideoHearingsService {
     getTelephoneConferenceId(hearingId: string): Observable<PhoneConferenceResponse> {
         return this.bhClient.getTelephoneConferenceIdById(hearingId);
     }
+
+    canAddUser(username: string): boolean {
+        const userPresent = this.modelHearing.participants.find(x => x.username.toLowerCase() === username.toLowerCase());
+        return !userPresent;
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/services/video-hearings.service.ts:436]:
```
return this.modelHearing.participants.every(x => x.username.toLowerCase() !== username.toLowerCase());
```

File: AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts [hunk 19]
```diff
@@ -182,73 +200,45 @@ export class AssignJudgeComponent extends BookingBaseComponent implements OnInit
...
+    get displayEmailField(): boolean {
+        return !!this.judge && this.judge.is_courtroom_account;
     }
 
     changeDisplayName() {
+        const text = SanitizeInputText(this.judgeDisplayNameFld.value);
+        this.judgeDisplayNameFld.setValue(text);
+
         if (this.judge && this.judge.display_name) {
+            // TODO confirm if this is redundant as set value will trigger these changes anyway
```
Comment (phillipjchambers) [inline AdminWebsite/AdminWebsite/ClientApp/src/app/booking/assign-judge/assign-judge.component.ts:235]: Remove comments

**Review by phillipjchambers — DISMISSED** [review]
I'm approving this, functionally looks good and changes are in line with the current approach/implementation. However, a lot of these components need refactoring & simplifying, to use standard angular approaches to the functionality.

**Review by mulvaneyjHMCTS — APPROVED** [review]

**Pull request merged**